docs_src/design/adr/0026-Common Configuration.md

```diff
@@ -0,0 +1,401 @@
+
+### Current Design
+
+The following flow chart demonstrates the bootstrapping of each services' configuration in the current Levski release.
+
+![](common-config-images/EdgeX 2.x Configuration bootstrapping flowchart.png)
+
+## Proposed Design
+The configuration settings that are common to all services will be partitioned out into a separate common configuration source. This  common configuration source will be pushed into the Configuration Provider by a new `core-common-config` service. When the Configuration Provider is not used the `core-common-config` service will provide the common configuration via a RESTful endpoint.
+
+During bootstrapping, each service will either load the common configuration from the Configuration Provider or via URI to some endpoint that provides the common configuration. This could be the new service's REST endpoint. Each service will have additional private configuration, which may overwrite and/or extend the common configuration.  
+
+An additional common configuration setting must be present to indicate all other common settings have been pushed to the Configuration Provider. This setting is stored last and the services must wait for this setting to be present prior to pulling the common settings.
+
+### Configuration Bootstrapping
+
+The following flow chart demonstrates the bootstrapping of each services' configuration with this new common configuration capability.
+
+![](common-config-images/EdgeX 3.x Configuration bootstrapping flowchart.png)
+
+### Secret Store Configuration
+
+As part of this design, the Secret Store configuration is being removed from the service configuration (common and private). This is so the Secret Provider can be instantiated prior to processing the service's configuration which may require the Secret Provider. The Secret Store configuration will now be a combination of default values and environment variable overrides. These environment variables will be the same as the ones that are currently used to override the configuration.
+
+### Specifying the Common Configuration location
+
+If the `-cp/--configProvider` command line option is used, the service will default to pulling the common configuration from a standard path in the Configuration Provider. i.e. `edgex/3.0/common/` The `-cp/--configProvider` option assumes the usage of the core-common-config service and cannot be used with the `-cc/--commonConfig` option.
+
+The new `-cc/--commonConfig` command line option will be added for all services. This option will take the URI that specifies where the common configuration is pulled when not using the Configuration Provider. If the `-cc/--commonConfig` option is not specified, the URI will default to the new `core-common-config` service's REST API. Authentication will be limited to `basic-auth`. In addition, a new environment override variable `EDGEX_COMMON_CONFIG` will be added which allows overriding this new command line option.

```

Comment:
```suggestion
The new `-cc/--commonConfig` command line option will be added for all services. This option will take the URI that specifies where the common configuration is pulled when not using the Configuration Provider. In addition, a new environment override variable `EDGEX_COMMON_CONFIG` will be added which allows overriding this new command line option. For `http`/`https` URI schemes, the authentication will be limited to Basic Authentication.

It is mandatory to set a URI for common configuration whenever Configuration Provider is not used.
```

Reply:
I forgot about the environment variable. 
So it seems as though you could use EDGEX_COMMON_CONFIG to set the location of the configuration and then not need to provide the `-cc/--commonConfig` command line option on every service? (Since the environment variable overrides anything set by `-cc/--commonConfig` anyway.)
Thus, EDGEX_COMMON_CONFIG becomes the "default" fallback that we couldn't figure out how to define.

Reply:
@TomBrennan-Eaton The environment variable EDGEX_COMMON_CONFIG is an **override** for what is provided as the command line flag using `-cc/--commonConfig`. EDGEX_COMMON_CONFIG will always take precedence whether `-cc` was used or not. Because EDGEX_COMMON_CONFIG is just a variable that can be set, it cannot be the default or fallback; it will always have the final say when it is set.

Reply:
@ejlee3 Agreed on how EDGEX_COMMON_CONFIG always overrides `-cc/--commonConfig`
Just pointing out that if `-cc/--commonConfig` isn't provided, EDGEX_COMMON_CONFIG can still be used, to good effect.

Reply:
@TomBrennan-Eaton the common environment variables always act this way; see https://docs.edgexfoundry.org/2.3/microservices/configuration/CommonEnvironmentVariables/